### PR TITLE
feat: Overlong filtering for GRPO

### PR DESCRIPTION
**Pull request by jubick1337**
- Add overlong_filtering flag to GRPOConfig
- Set loss_multiplier to 0 for truncated samples when flag is enabled
- Set token_loss_mask to 0 for assistant tokens in truncated samples
- Add truncated field to batch in both sync and async rollout functions

# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/algorithms/grpo.py [hunk 1]
```diff
@@ -604,13 +605,33 @@ def grpo_train(
                     )
 
             with timer.time("data_processing"):
+                use_overlong_filtering = master_config["grpo"].get(
```
Comment (SahilJain314) [inline nemo_rl/algorithms/grpo.py]: We've banned using default dict args inside of the codebase (we should probably write this fact down somewhere @terrykong). Can you switch this to just call the arg directly? If a config is missing this key, it should crash.

File: nemo_rl/algorithms/grpo.py [hunk 2]
```diff
@@ -604,13 +605,33 @@ def grpo_train(
...
                 for i, message_log in enumerate(repeated_batch["message_log"]):
+                    overlong_filtered = (
+                        use_overlong_filtering and repeated_batch["truncated"][i]
+                    )
                     for j, message in enumerate(message_log):
                         if message["role"] == "assistant":
-                            message["token_loss_mask"] = torch.ones_like(
-                                message["token_ids"]
-                            )
+                            if overlong_filtered:
```
Comment (SahilJain314) [inline nemo_rl/algorithms/grpo.py]: this part of the logic is not required. Just setting the loss_multiplier is sufficient (corresponds to sample_mask) in the loss function

File: nemo_rl/algorithms/grpo.py (hunk 1 above)
Comment (terrykong) [reply nemo_rl/algorithms/grpo.py]: yea, good point. i've opened https://github.com/NVIDIA-NeMo/RL/pull/810/files

**Review by ashors1 — APPROVED** [review]
Thank you for the PR @jubick1337! This looks good to me. @SahilJain314 any other comments from your side?

**Comment by jubick1337** [comment]
Hi @SahilJain314, gentle ping when you have a moment to take another look. Thanks!

File: docs/guides/grpo.md [hunk 3]
```diff
@@ -187,6 +187,30 @@ By multiplying the first term of the loss function by the importance weights $\f
...
 
+#### Overlong Filtering (overlong_filtering)
+
+When `overlong_filtering=True`, samples that reach `max_total_sequence_length` without producing an end-of-text token are excluded from loss computation. This reduces noise from penalizing generations that may be high-quality but exceed the sequence length limit.
+
+The implementation modifies the loss calculation as follows:
+
+For each sample $i$ in the batch:
+- Let $\text{truncated}_i = \begin{cases} 1 & \text{if sample } i \text{ reached max length without EOS} \\ 0 & \text{otherwise} \end{cases}$
+- The sample mask becomes: $\text{sample\_mask}_i = \text{loss\_multiplier}_i \cdot (1 - \text{truncated}_i)$
```
Comment (ashors1) [inline docs/guides/grpo.md]: This isn't rendering properly. Can you take a look?

Comment (jubick1337) [reply docs/guides/grpo.md]: @ashors1, thanks for spotting this! I updated this section.

File: nemo_rl/algorithms/grpo.py [hunk 4]
```diff
@@ -626,6 +627,16 @@ def grpo_train(
                     )
 
             with timer.time("data_processing"):
+                use_overlong_filtering = master_config["grpo"]["overlong_filtering"]
+                if use_overlong_filtering and "truncated" in repeated_batch:
```
Comment (ashors1) [inline nemo_rl/algorithms/grpo.py]: Do you think we should error out if `use_overlong_filtering` is True but `truncated` is not in the batch?

Comment (jubick1337) [reply nemo_rl/algorithms/grpo.py]: Thanks for the review! I've removed the defensive check entirely.

Since the `truncated` field is always added by our rollout functions (both sync and async) as part of this PR, it should always be present.

**Review by ashors1 — DISMISSED** [review]

**Review by terrykong — APPROVED** [review]

**Comment by ashors1** [comment]
Hi @jubick1337, the presubmit CI failed because new multimodal recipes were recently added to main: https://github.com/NVIDIA-NeMo/RL/actions/runs/17089287317/job/48467257968#step:3:7379. Would you be able to help resolve this? Then we can merge

**Review by terrykong — DISMISSED** [review]

**Comment by jubick1337** [comment]
@terrykong I updated vlm configs, should be good now
